### PR TITLE
DocC Version Switcher

### LINKED ISSUE #1834: Include a version switcher in the DocC header breadcrumb
We're slightly changing the rules on the breadcrumb here and making it active navigation, but with the changes alerting people to being on a non-stable docs version, I think this will work OK.

The dropdown may need custom styling to not look 🤮  but I want to avoid much (if not all) JavaScript here so we don't run the risk of interfering with Vue that powers the DocC section of the page.

Image

### PR DESCRIPTION
**Pull request by daveverwer**
Fixes #1834 

Feature gated for development/staging at the moment until the other tickets are also implemented.

Screenshot 2022-07-18 at 19 29 31@2x

### CODE REVIEW

**Comment by Sherlouk** [comment]
how exciting

File: Sources/App/Controllers/PackageController+routes.swift [hunk 1]
```diff
@@ -103,20 +112,65 @@ struct PackageController {
...
                     .field(Version.self, \.$packageName)
-                    .field(Version.self, \.$spiManifest)
+                    .field(Version.self, \.$docArchives)
                     .field(Repository.self, \.$ownerName)
-                    .first(),
-                      let body = awsResponse.body,
+                    .all()
+
+                let documentationVersions = queryResult.compactMap { result -> DocumentationVersion? in
+                    guard let reference = result.model.reference else { return nil }
```
Comment (finestructure) [inline Sources/App/Controllers/PackageController+routes.swift:130]: Getting a warning here - `result.model.reference` isn't actually optional.

File: Sources/App/Controllers/PackageController+routes.swift [hunk 2]
```diff
@@ -103,20 +112,65 @@ struct PackageController {
...
+                }
+
+                let availableDocumentationVersions: [DocumentationPageProcessor.AvailableDocumentationVersion] = ([
+                    documentationVersions.filter { $0.latest == .defaultBranch }.first,
+                    documentationVersions.filter { $0.latest == .preRelease }.first
+                ] + documentationVersions.latestMajorVersions())
+                    .compactMap { version in
+                        guard let version = version,
+                              let latest = version.latest,
+                              let docArchives = version.docArchives
```
Comment (finestructure) [inline Sources/App/Controllers/PackageController+routes.swift:157]: Getting a warning here - `version.docArchives` isn't actually optional.

File: Sources/App/Controllers/PackageController+routes.swift [hunk 3]
```diff
@@ -103,20 +112,65 @@ struct PackageController {
...
+
+                let documentationVersions = queryResult.compactMap { result -> DocumentationVersion? in
+                    guard let reference = result.model.reference else { return nil }
+
+                    return DocumentationVersion(reference: reference,
+                                                ownerName: result.relation2?.ownerName ?? owner,
+                                                packageName: result.model.packageName ?? repository,
+                                                docArchives: (result.model.docArchives ?? []).map(\.title),
+                                                latest: result.model.latest)
+                }
```
Comment (finestructure) [inline Sources/App/Controllers/PackageController+routes.swift:137]:
```suggestion
                let documentationVersions = queryResult.compactMap { result -> DocumentationVersion? in
                    DocumentationVersion(reference: reference,
                                                ownerName: result.relation2?.ownerName ?? owner,
                                                packageName: result.model.packageName ?? repository,
                                                docArchives: (result.model.docArchives ?? []).map(\.title),
                                                latest: result.model.latest)
                }
```

File: Sources/App/Controllers/PackageController+routes.swift [hunk 4]
```diff
@@ -103,20 +112,65 @@ struct PackageController {
...
+                    documentationVersions.filter { $0.latest == .defaultBranch }.first,
+                    documentationVersions.filter { $0.latest == .preRelease }.first
+                ] + documentationVersions.latestMajorVersions())
+                    .compactMap { version in
+                        guard let version = version,
+                              let latest = version.latest,
+                              let docArchives = version.docArchives
+                        else { return nil }
+
+                        return .init(kind: latest, reference: "\(version.reference)", docArchives: docArchives)
```
Comment (finestructure) [inline Sources/App/Controllers/PackageController+routes.swift:160]:
```suggestion
                        guard let version = version,
                              let latest = version.latest
                        else { return nil }

                        return .init(kind: latest, reference: "\(version.reference)", docArchives: version.docArchives)
```

File: Sources/App/Controllers/PackageController+routes.swift [hunk 5]
```diff
@@ -304,3 +358,37 @@ private extension HTTPHeaders {
...
+            guard let semVer = version.reference.semVer else { return false }
+            return semVer.isStable
+        }
+        let groupedStableVersions = Dictionary.init(grouping: stableVersions) { version in
+            version.reference.semVer?.major
+        }
+
+        return groupedStableVersions.compactMap { key, versions -> Element? in
+            // If any of the references had a nil semVer then there could be a nil key in the dictionary.
+            guard let _ = key else { return nil }
```
Comment (finestructure) [inline Sources/App/Controllers/PackageController+routes.swift:378]: `guard key != nil else { return nil }` ?

or perhaps

```swift
        let groupedStableVersions = Dictionary.init(grouping: stableVersions) { version in
             version.reference.semVer?.major
         }
        .compactMap { $0.key }
```

**Review by finestructure — APPROVED** [review]

**Comment by finestructure** [comment]
Beautiful!

File: Sources/App/Controllers/PackageController+routes.swift (hunk 4 above)
Comment (finestructure) [reply Sources/App/Controllers/PackageController+routes.swift:160]: Looks like that's also breaking the tests (compile error) - but how is that not failing for you locally? 🤔 

https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/runs/7395781849?check_suite_focus=true

File: Sources/App/Controllers/PackageController+routes.swift [hunk 6]
```diff
@@ -304,3 +358,37 @@ private extension HTTPHeaders {
...
+
+            // Filter down to only the largest semVer in each group.
+            let latestMajorStableVersion = versions
+                .compactMap { result -> (result: Element, semVer: SemanticVersion)? in
+                    guard let semVer = result.reference.semVer else { return nil }
+                    return (result: result, semVer: semVer)
+                }
+                .sorted(using: KeyPathComparator(\.semVer, order: .reverse))
```
Comment (finestructure) [inline Sources/App/Controllers/PackageController+routes.swift:386]: `KeyPathComparator` is not available on Linux it seems:

```
❌ /__w/SwiftPackageIndex-Server/SwiftPackageIndex-Server/Sources/App/Controllers/PackageController+routes.swift:392:24: cannot find 'KeyPathComparator' in scope
        .sorted(using: KeyPathComparator(\.reference.semVer))
```

Comment (daveverwer) [reply Sources/App/Controllers/PackageController+routes.swift:386]: Ahh. I saw the tests failing last night with valid errors, but the tests passed on my local machine. I couldn't spend another minute in the heat of the office, though! I'll swap these out for regular sorted.

It's a shame, it's a really nice API.

File: Sources/App/Controllers/PackageController+routes.swift (hunk 1 above)
Comment (daveverwer) [reply Sources/App/Controllers/PackageController+routes.swift:130]: This is so odd. I wasn't getting these on Xcode 14 at all. Switched back to 13 now, and they show up.

File: Sources/App/Controllers/PackageController+routes.swift (hunk 2 above)
Comment (daveverwer) [reply Sources/App/Controllers/PackageController+routes.swift:157]: Same here

**Comment by daveverwer** [comment]
One last visual change before merging. I had removed the "Switch to…" from the server-side HTML so it was easier to check for empty arrays as an indicator of no other versions, intending to put it back in CSS but had never got around to doing that. It now looks like this:

Screenshot 2022-07-19 at 12 37 05@2x

**Pull request merged**